ADR for extra tempo receivers (#185)
canonical/observability: decision-records/2024-06-25--extra-tempo-receivers.md

However, that is only applicable to **charmed workloads**. For uncharmed applications, integration is through grafana agent, as specified in https://discourse.charmhub.io/t/how-to-integrate-cos-lite-with-uncharmed-applications/12005.
Since requesting tracing endpoints from tempo charm is done through charm relations, there needs to be another way to tell tempo to provide an endpoint with a specific protocol to the uncharmed application that wishes to push traces to tempo even when there are no relations requesting that protocol.

## Decision

### Config option for each receiver
Have multiple config options to enable a receiver for each protocol we support for tempo charm. These **extra** receivers will tell tempo to provide endpoints for those protocols and will be unioned with those coming from tracing relations so that uncharmed applications can send traces using any supported protocol without depending on relation data.
```

> juju config tempo enable_zipkin=True

Reviewer: I like how `always_enable` sounds like in this context.

> juju config tempo enable_zipkin=True

Reviewer: This still needs to be updated, right @michaeldmitry?

> juju config tempo enable_zipkin=True

Author: Yeah I think I'll go with `always_enable` if we all agree on that.

juju config tempo enable_kafka=True
juju config tempo enable_opencensus=True
juju config tempo enable_tempo_http=True
juju config tempo enable_tempo_grpc=True
juju config tempo enable_otlp_grpc=True
juju config tempo enable_otlp_http=True
```
## Benefits
- Safeguarding against typos, unsupported protocols, duplications, ...etc. (e.g: otlp_https)
- We don't have to worry about storing the extra receivers somewhere.
- Simplicity in UX.

> - Set of protocols is limited, and hopefully they will be reduced even more in the future.

Reviewer 2: ```suggestion
```

## Disadvantages

> - Maintenance overhead, as config options are dependable on the current supported protocols.

Reviewer 2: ```suggestion
- Maintenance: config options depend on the currently supported protocols, there is a chance that future versions of tempo may add/remove some and force the charm's API to change in non-backwards-compatible ways. Hopefully, however, the set of supported protocols will only shrink as the industry converges towards a smaller set of standards (otlp?)
- Ugliness: having individual flags for each enum member to tag a subset of it is not a very neat design, but ATM it's all that Juju offers us. See [the charmhub page for mimir-worker](https://charmhub.io/mimir-worker-k8s/configuration) to see what it looks like in practice. The UX is also not super: you can only enable/disable one at a time.
```

## Alternatives considered

### Comma separated config
Have a single config option to provide a comma-separated list of roles.

```
juju config tempo receivers=oltp_grpc,otlp_http
```
## Benefits
- We don't have to worry about storing the extra receivers somewhere.
## Disadvantages
- Prone to typos, unsupported protocols, duplications, ...etc. (e.g: otlp_https)
- Poor UX.

### Actions
Set extra receivers using an action to enable a specific extra receiver.
```
juju run tempo/0 enable-receiver otlp_grpc
```
## Benefits

> - Safeguarding against typos, unsupported protocols, duplications, ...etc. (e.g: otlp_https)

Reviewer 2: we don't need to give an example every time we mention typos :)

## Disadvantages
- We have to store the extra receivers somewhere (e.g: peer relation data).

> - Actions run on units, not apps, and are typically run on the leader unit.

Reviewer 2: ```suggestion
- Actions run on units, not apps. Hard to manage user expectations as to what'd happen if you run it on a follower unit vs. the leader.
```

# Extra tempo receivers
**Date:** 2024-06-25<br/>
**Authors:** @PietroPasotti, @mmkay, @michaeldmitry

## Context and Problem Statement
Charms seeking to push traces to Tempo will relate to the Tempo charm over a `tracing` relation. [The protocol](https://github.com/canonical/charm-relation-interfaces/tree/main/interfaces/tracing/v2) specifies that the requirer charm needs to request a list of the tracing protocols it intends to use to send spans to Tempo (e.g: `otlp_http`, `jaeger`). Then, as a response, the provider Tempo will share for each requested protocol (so long as it supports it), an endpoint to which workloads/charms can push traces.

> Tempo only enables receivers for the protocols it actually needs to satisfy all incoming requests. Therefore, if a user wants to send traces to Tempo over a protocol that no charm is currently requesting via the `tracing` interface, there needs to be a way to configure the Tempo charm to enable that receiver regardless of whether there is a `tracing` integration asking for it the regular way. 

Reviewer: It is also our choice (although a rather obvious one) to support uncharmed tracing. We might want to mention that we decided to allow uncharmed applications to send traces as well.

The choice now is: how do we allow the end user of the Tempo charm to specify which subset of the possible tracing protocols to always enable, regardless of integrations? 

## Decision

### Config option for each receiver
#### Syntax:
The decision is, for each protocol Tempo supports, to expose a juju config option on the tempo-coordinator charm to force-enable it. 
Thus for each supported protocol `X` there will be an `always_enable_X:bool` config option.
#### Semantics:
The coordinator charm will collect all protocols that are configured to "always_enable" and add those to the set of protocols that are requested through active tracing relations from charmed workloads.
```
juju config tempo enable_zipkin=True
juju config tempo enable_kafka=True

> juju config tempo enable_opencensus=True

Reviewer: Minor nit: we likely don't need kafka or opencensus as they're not allowed values now.

juju config tempo enable_tempo_http=True
juju config tempo enable_tempo_grpc=True
juju config tempo enable_otlp_grpc=True
juju config tempo enable_otlp_http=True
```

> ## Benefits

Reviewer: Another nit: we likely need it to be 
```suggestion
## Benefits
#### Benefits
``` 

to render correctly. Same change needs to be applied in all "Benefits" and "Disadvantages" headers below.

- We don't have to worry about storing the extra receivers somewhere, as config is persisted in Juju.
- Consistency with [the enum pattern implemented for mimir roles](https://github.com/canonical/mimir-worker-k8s-operator/blob/main/config.yaml).
- Simplicity in UX.
## Disadvantages
- Maintenance: config options depend on the currently supported protocols, there is a chance that future versions of tempo may add/remove some and force the charm's API to change in non-backwards-compatible ways. Hopefully, however, the set of supported protocols will only shrink as the industry converges towards a smaller set of standards (otlp?)
- Ugliness: having individual flags for each enum member to tag a subset of it is not a very neat design, but ATM it's all that Juju offers us. See [the charmhub page for mimir-worker](https://charmhub.io/mimir-worker-k8s/configuration) to see what it looks like in practice. The UX is also not super: you can only enable/disable one at a time.

## Alternatives considered

### Comma separated config
Have a single config option to provide a comma-separated list of roles.

```

> juju config tempo receivers=oltp_grpc,otlp_http

Reviewer: nit:

```suggestion
juju config tempo receivers=oltp_grpc,otlp_http
juju config tempo receivers=otlp_grpc,otlp_http
```

Reviewer 2: hahaha a perfect example of why we didn't go this way :D 
I say let's keep it
